docs/ssl/gen_certs.rb

```diff
@@ -0,0 +1,175 @@
+    @cert.issuer = @cert.subject # root CA's are "self-signed"
+    @cert.public_key = @key.public_key
+    @cert.not_before = Time.now
+    @cert.not_after = @cert.not_before + 2 * 365 * 24 * 60 * 60 # 2 years
+    ef = OpenSSL::X509::ExtensionFactory.new
+    ef.subject_certificate = @cert
+    ef.issuer_certificate = @cert
+    @cert.add_extension(ef.create_extension("basicConstraints","CA:TRUE",true))
+    @cert.add_extension(ef.create_extension("keyUsage","keyCertSign, cRLSign", true))
+    @cert.add_extension(ef.create_extension("subjectKeyIdentifier","hash",false))
+    @cert.add_extension(ef.create_extension("authorityKeyIdentifier","keyid:always",false))
+    #@cert.sign(@key, OpenSSL::Digest::SHA256.new)
+  end
+
+  # literal objects for Ruby internal use
+  attr_reader :key
+  attr_reader :cert
+
+  # need string representations for use in files
+  def key_pem
+    @key.to_pem('aes256', @@ca_pass)
+  end
+
+  def cert_pem
+    @cert.to_pem
+  end
+
+  # used by the server cert to keep serials increasing
+  def assign_serial
+    @@ca_serial += 1
+  end
+
+  def _get_parent_certs
+    r = []
+    if @ca
+      r = @ca.get_parent_certs
+    end
+    r.push( @cert )
+    r
+  end
+
+  # return the cert store object containing this CA and all parents
+  def cert_chain
+    cert_store = OpenSSL::X509::Store.new
+    for cert in _get_parent_certs
+      cert_store.add_cert cert
+    end
+    cert_store
+  end
+
+  # return the same chain as cert_chain, but in a PEM string format
+  def cert_chain_pem
+    r = ""
+    for cert in _get_parent_certs
+      r += cert.to_pem
+    end
+    r
+  end
+end
+
+class RootCA < CA
+
+  def initialize(bits=4096, subject="C=US, ST=IL, O=Danny Sauer dot com, OU=Danny Sauer dot com Certificate Authority, CN=Danny Sauer dot com Root CA/emailAddress=danny@dannysauer.com")
+    super(bits, subject)
+    @cert.sign(@key, OpenSSL::Digest::SHA256.new)
+  end
+
+  
+end
+
+class IntermediateCA < CA
+  @@ilevel = 0
+
+  def initialize(ca, bits=4096, subject="C=US, ST=IL, O=Danny Sauer dot com, OU=Danny Sauer dot com Certificate Authority, CN=Danny Sauer dot com Intermediate CA level %s/emailAddress=danny@dannysauer.com" % (@@ilevel+=1))
+    @ca = ca
+    super(bits, subject)
+    @cert.sign(@ca.key, OpenSSL::Digest::SHA256.new)
+  end
+end
+
+class ServerCert
+  def initialize(
+    ca,
+    altname_list=nil,
+    start_date=Time.now,  # note that Ruby seems to refuse to create a cert w/ no date
+    lifetime_days=1 * 365 * 24 * 60 * 60 # 1 year validity
+  )
+    @ca = ca
+
+    @key  = OpenSSL::PKey::RSA.new 2048
+    @cert = OpenSSL::X509::Certificate.new
+    @cert.version = 2 # must be 2
+    @cert.serial = @ca.assign_serial
+    @cert.subject = OpenSSL::X509::Name.parse "/DC=org/DC=ruby-lang/CN=Ruby certificate"
+    @cert.issuer = @ca.cert.subject # CA is the issuer
+    @cert.public_key = @key.public_key
+    if start_date
+      @cert.not_before = start_date
+    end
+    if lifetime_days
```

Comment:
It seems to me that since start date is required, that lifetime days should be required as well. ( at least it cannot be nil ) Defaulting to 1 days is fine, but it should always be something.